File: java/src/main/java/cn/zenliu/java/translator/StreamMapDictFactory.java
package cn.zenliu.java.translator;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public final class StreamMapDictFactory {
    StreamMapDictFactory() {
    }

    static final char KEY_VALUE_SPLICER = ' ';
    static final char VALUES_SPLICER = '\'';

    static class StreamDict extends MapDictionary {
        StreamDict(final InputStream txtStream) {
            this.map = new BufferedReader(new InputStreamReader(txtStream))
                    .lines()
                    .parallel()
                    .map(l -> {
                        final String[] split = l.split(String.valueOf(KEY_VALUE_SPLICER));
                        if (split.length != 2)
                            return null;
                        else
                            return split;
                    })
                    .filter(Objects::nonNull)
                    .collect(Collectors.toMap(
                            s -> s[0],
                            s -> Arrays.stream(s[1].split(String.valueOf(VALUES_SPLICER)))
                                    .parallel()
                                    .filter(it -> !it.isEmpty())
                                    .collect(Collectors.toList()).toArray(new String[0])
                            )
                    );

        }

        final Map<String, String[]> map;

        Map<String, String[]> getMapping() {
            return map;
        }
    }

    public static Dictionary build(InputStream textStream) {
        return new StreamDict(textStream);
    }
}

File: kotlin/src/main/kotlin/cn/zenliu/kotlin/translator/StreamMapDictFactory.kt
package cn.zenliu.kotlin.translator

import java.io.*
import kotlin.streams.*

object StreamMapDictFactory {
	class StreamDict internal constructor(textInputStream: InputStream) : MapDictionary() {
		override val mapping: Map<String, Array<String>> = textInputStream
			.let { BufferedReader(InputStreamReader(it)) }
			.lines()
			.map {
				it
					.split(KEY_VALUE_SPLITER)
					.dropLastWhile { it.isEmpty() }
					.toTypedArray()
					.takeIf { it.size == 2 }
					?.let {
						it[1] to it[0].split(VALUES_SPLITER).dropLastWhile { it.isNullOrBlank() }.toTypedArray()
					}
			}
			.toList()
			.filterNotNull()
			.toMap()
	}

	var KEY_VALUE_SPLITER = ' '
	var VALUES_SPLITER = '\''
	/**
	 * build dictionary from format inputstream
	 * @param textInputStream InputStream
	 * @return StreamDict
	 */
	fun buildDict(textInputStream: InputStream) = StreamDict(textInputStream)

}
